src/core/infer.ml

```diff
@@ -14,7 +14,7 @@ let constraints = Stdlib.ref []
 
 (** Function adding a constraint. *)
 let conv ctx a b =
-  if not (Basics.eq ctx a b) then
+  if not (Eval.eq_modulo ctx a b) then
```

Comment:
Using equality here had a significant positive impact on performances if I remember well. Conversion on the other end can be very costly, and if it fails it is gonna be tested again partially by unification.

Reply:
This is not clear. We should compare the performances.

Reply:
We can't compare anymore because `eq` is in the wrong module now (cyclic depencency).